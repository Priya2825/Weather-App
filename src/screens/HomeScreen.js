import React, { useEffect, useState } from 'react';
import {
  View, FlatList, SafeAreaView, StyleSheet,
  TextInput, Text, ActivityIndicator,
} from 'react-native';
import WeatherCard from '../../components/WeatherCard';
import API from '../../Utils/API';
import axios from 'axios';
import Icon from 'react-native-vector-icons/Ionicons';
import * as Notifications from 'expo-notifications';
import AsyncStorage from '@react-native-async-storage/async-storage';

function HomeScreen({ navigation }) {
  const [weatherData, setWeatherData] = useState([]);
  const [searchText, setSearchText] = useState('');
  const [filteredData, setFilteredData] = useState([]);
  const [coordinates, setCoordinates] = useState([]);
  const [searchcoordinates, setSearchCoordinates] = useState();
  const [errorText, setErrorText] = useState(null);
  const [isloading, setIsloading] = useState(false);

  useEffect(() => {
    if (coordinates.length > 0) {
      GetWeatherData(coordinates);
    } else {
      setWeatherData([]);
      setFilteredData([]);
      setIsloading(false);
      setErrorText(null);
    }
  }, [coordinates]);

  useEffect(() => {
    loadCoordinates();
  }, []);

  useEffect(() => {
    if (searchText) {
      AdditionalCountryWeatherData();
    } else {
      setFilteredData(weatherData);
    }
  }, [searchText, weatherData]);

  function addLocation() {
    saveCoordinates(searchcoordinates);
  }

  function removeLocation(coord, name) {
    removeCoordinates(coord.lat, coord.lon, name);
  }

  function AdditionalCountryWeatherData() {
    setErrorText(null);
    setIsloading(true);
    const filtered = weatherData.filter((item) =>
      item.name.toLowerCase().includes(searchText.toLowerCase())
    );
    if (filtered.length > 0) {
      setIsloading(false);
      setFilteredData(filtered);
      return;
    }
    API.get('/weather', {
      params: {
        q: searchText.toLowerCase(),
        units: 'metric',
      },
    })
      .then((result) => {
        const { lat, lon } = result.data.coord;
        setSearchCoordinates({ lat, lon, name: result.data.name });
        setFilteredData([result.data]);
        setIsloading(false);
      })
      .catch((error) => {
        setFilteredData([]);
        setErrorText('Country not found');
        setIsloading(false);
      });
  }

  const loadCoordinates = async () => {
    try {
      const jsonValue = await AsyncStorage.getItem('coordinates');
      const loadedCoordinates = jsonValue != null ? JSON.parse(jsonValue) : [];
      setCoordinates(loadedCoordinates);
    } catch (error) {
      console.log('Error loading coordinates:', error);
    }
  };

  const saveCoordinates = async (newCoordinate) => {
    try {
      if (
        coordinates.filter((item) => item.name == newCoordinate.name).length > 0
      ) {
        return;
      }
      const updatedCoordinates = [...coordinates, newCoordinate];
      setCoordinates(updatedCoordinates);
      const jsonValue = JSON.stringify(updatedCoordinates);
      await AsyncStorage.setItem('coordinates', jsonValue);
    } catch (error) {
      console.log('Error saving coordinates:', error);
    }
  };

  const removeCoordinates = async (lat, lon, name) => {
    try {
      const updatedCoordinates = coordinates.filter(
        (item) => item.name != name
      );
      setCoordinates(updatedCoordinates);
      const jsonValue = JSON.stringify(updatedCoordinates);
      await AsyncStorage.setItem('coordinates', jsonValue);
    } catch (error) {
      console.log('Error removing coordinates:', error);
    }
  };

  function GetWeatherData(coordinateList) {
    setIsloading(true);
    const weatherRequests = coordinateList.map((item) =>
      API.get('/weather', {
        params: {
          lat: item.lat,
          lon: item.lon,
          units: 'metric',
        },
      })
    );

    axios
      .all(weatherRequests)
      .then(
        axios.spread((...result) => {
          const data = result.map((item) => item.data);
          setWeatherData(data);
          setFilteredData(data); // Initialize filtered data
          setIsloading(false);
        })
      )
      .catch((error) => {
        console.log(error);
        setIsloading(false);
      });
  }

  async function schedulePushNotification(title, body) {
    console.log(title, body);
    const value = await AsyncStorage.getItem('notification-enabled');
    if (value == 'true') {
      await Notifications.scheduleNotificationAsync({
        content: {
          title: title,
          body: body,
        },
        trigger: null,
      });
    }
  }

  return (
    <View style={styles.container}>
      <SafeAreaView>
        <View style={styles.searchContainer}>
          <Icon name="search" size={20} color="#555" style={styles.searchIcon} />
          <TextInput
            style={styles.searchBar}
            placeholder="Search and Add Cities"
            value={searchText}
            onChangeText={(text) => setSearchText(text)}
          />
        </View>
        {isloading ? (
          <View style={styles.loadingContainer}>
            <ActivityIndicator size="large" color="#6200ea" />
          </View>
        ) : errorText ? (
          <Text style={styles.errorText}>{errorText}</Text>
        ) : (
          <FlatList
            ListEmptyComponent={
              <View style={styles.loadingContainer}>
                <Text style={styles.emptyText}>No Cities Added! Please Add City!</Text>
              </View>
            }
            data={filteredData}
            renderItem={({ item }) => (
              <WeatherCard
                addLocation={addLocation}
                removeLocation={() => removeLocation(item.coord, item.name)}
                isAdded={
                  coordinates.find(
                    (coord) =>
                      coord.lat === item.coord.lat &&
                      coord.lon === item.coord.lon
                  ) !== undefined
                }
                location={item.name}
                temperature={item.main.temp}
                currentWeather={item.weather[0].main}
                handleOnPress={async (color, weatherIcon) => {
                  await schedulePushNotification(
                    `Weather Update: ${item.main.temp}Â°C`,
                    `Current weather is ${item.weather[0].description} in ${item.name}`
                  );
                  navigation.navigate('Detail', {
                    backgroundColor: color,
                    weatherIcon: weatherIcon,
                    ...item,
                  });
                }}
              />
            )}
            keyExtractor={(item, index) => index.toString()}
          />
        )}
      </SafeAreaView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 12,
    backgroundColor: '#F3F4F6',
  },
  loadingContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  searchContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    borderColor: '#ddd',
    borderWidth: 1,
    borderRadius: 10,
    paddingHorizontal: 12,
    paddingVertical: 8,
    marginBottom: 12,
    backgroundColor: '#FFF',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 3,
  },
  searchIcon: {
    marginRight: 10,
  },
  searchBar: {
    flex: 1,
    height: 40,
    color: '#333',
  },
  errorText: {
    color: '#D32F2F',
    textAlign: 'center',
    marginTop: 20,
    fontSize: 16,
  },
  emptyText: {
    color: '#757575',
    fontSize: 16,
  },
});

export default HomeScreen;
