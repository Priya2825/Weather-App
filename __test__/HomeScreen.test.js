import React from "react";
import HomeScreen from "../src/screens/HomeScreen"; // Adjust the path as needed
import AsyncStorage from "@react-native-async-storage/async-storage";
import renderer from "react-test-renderer";
import {
  render,
  waitFor,
  fireEvent,
  screen,
} from "@testing-library/react-native";
import api from "../Utils/API";
// import api from "../Utils/API";

// Mock AsyncStorage and axios
jest.mock("@react-native-async-storage/async-storage", () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

jest.mock("../Utils/API");

describe("HomeScreen", () => {
  afterEach(() => {
    jest.clearAllMocks(); // Clear mocks between tests
  });

  it("renders correctly snapshot", async () => {
    const tree = renderer.create(<HomeScreen />).toJSON();
    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
  });

  it("should render correctly", async () => {
    // Mock AsyncStorage to return no saved coordinates
    AsyncStorage.getItem.mockResolvedValue(null);

    const { getByPlaceholderText, getByText } = render(<HomeScreen />);

    // Wait for state updates and effects to complete
    await waitFor(() => {
      // Check if the search bar is rendered
      expect(getByPlaceholderText("Search and Add Cities")).toBeTruthy();

      // Check if some placeholder text or a default element is rendered, like a message or empty list
      expect(getByText("No Cities Added! Please Add City!")).toBeTruthy(); // Assuming you have this text when there's no data
    });
  });

  it("should render saved cities from AsyncStorage", async () => {
    // Mock saved city data
    const savedCities = JSON.stringify([
      { lat: 10, lon: 10, name: "City1" },
      { lat: 20, lon: 20, name: "City2" },
    ]);

    // Mock AsyncStorage to return saved cities
    AsyncStorage.getItem.mockResolvedValue(savedCities);

    // Mock API responses
    api.get.mockImplementation((url, { params }) => {
      if (url === "/weather") {
        if (params.lat === 10 && params.lon === 10) {
          return Promise.resolve({
            data: {
              coord: { lat: 10, lon: 10 },
              name: "City1",
              main: { temp: 25 },
              weather: [{ main: "Sunny", description: "clear sky" }],
            },
          });
        } else if (params.lat === 20 && params.lon === 20) {
          return Promise.resolve({
            data: {
              coord: { lat: 20, lon: 20 },
              name: "City2",
              main: { temp: 30 },
              weather: [{ main: "Cloudy", description: "overcast clouds" }],
            },
          });
        }
        return Promise.reject(new Error("Unknown coordinates"));
      }
      return Promise.reject(new Error("Invalid URL"));
    });

    const { getByText } = render(<HomeScreen />);

    await waitFor(() => {
      // Verify that weather data is rendered
      expect(getByText("City1")).toBeTruthy();
      expect(getByText("25째C")).toBeTruthy();
      expect(getByText("Sunny")).toBeTruthy();

      expect(getByText("City2")).toBeTruthy();
      expect(getByText("30째C")).toBeTruthy();
      expect(getByText("Cloudy")).toBeTruthy();
    });
  });

  //   // 2. Test adding a city by mocking an API call and rendering the city
  //   it("should add a city and render it", async () => {
  //     // Mock an empty AsyncStorage
  //     AsyncStorage.getItem.mockResolvedValue(null);

  //     // Mock axios response for the API call
  //     api.get.mockResolvedValue({
  //       data: {
  //         coord: { lat: 1.4002, lon: 103.8147 },
  //         name: "London",
  //         main: { temp: 25 },
  //         weather: [{ main: "Cloudy", description: "overcast clouds" }],
  //       },
  //     });

  //     const { getByPlaceholderText, getByText } = render(<HomeScreen />);

  //     // Simulate entering a city name and adding it
  //     const input = getByPlaceholderText("Search and Add Cities");
  //     fireEvent.changeText(input, "London");

  //     // Simulate API call to add city (e.g., on button press)
  //     fireEvent.press(getByText("Add")); // Assuming there's a button with this label

  //     // Wait for the city to be added and rendered
  //     await waitFor(() => {
  //       expect(getByText("London")).toBeTruthy();
  //       expect(getByText("25째C")).toBeTruthy(); // Assuming temperature is displayed
  //       expect(getByText("Cloudy")).toBeTruthy();
  //     });
  //   });

  it("should add a city and render it", async () => {
    // Mock AsyncStorage with no saved cities initially
    AsyncStorage.getItem.mockResolvedValueOnce(JSON.stringify([]));

    // Mock API call for London
    api.get.mockResolvedValue({
      data: {
        coord: { lat: 20, lon: 20 },
        name: "London",
        main: { temp: 30 },
        weather: [{ main: "Cloudy", description: "overcast clouds" }],
      },
    });

    // Render the HomeScreen
    const { getByPlaceholderText, getByText, queryByText, debug } = render(
      <HomeScreen />
    );

    // Simulate search input
    const searchInput = getByPlaceholderText("Search and Add Cities");
    fireEvent.changeText(searchInput, "London");

    // Wait for any error messages to ensure no errors
    await waitFor(() => {
      expect(queryByText("Country not found")).toBeNull();
    });

    // Simulate pressing the "Add" button to add the city
    const addButton = getByText("Add"); // Assuming the "Add" button exists after search
    fireEvent.press(addButton);

    // Mock AsyncStorage to return the added city after the button press
    AsyncStorage.getItem.mockResolvedValueOnce(
      JSON.stringify([{ lat: 20, lon: 20, name: "London" }])
    );

    // Wait for the city to be rendered in the list
    await waitFor(() => {
      expect(getByText("London")).toBeTruthy(); // City name
      expect(getByText("30째C")).toBeTruthy(); // Assuming temperature display
      expect(getByText("Cloudy")).toBeTruthy(); // Assuming weather description
    });
  });
});
