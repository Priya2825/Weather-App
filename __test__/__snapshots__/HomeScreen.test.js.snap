// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#F3F4F6",
      "flex": 1,
      "padding": 12,
    }
  }
>
  <RCTSafeAreaView>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFF",
          "borderColor": "#ddd",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginBottom": 12,
          "paddingHorizontal": 12,
          "paddingVertical": 8,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 3,
        }
      }
    >
      <Text />
      <TextInput
        onChangeText={[Function]}
        placeholder="Search and Add Cities"
        style={
          {
            "color": "#333",
            "flex": 1,
            "height": 40,
          }
        }
        value=""
      />
    </View>
    <RCTScrollView
      ListEmptyComponent={
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "#757575",
                "fontSize": 16,
              }
            }
          >
            No Cities Added! Please Add City!
          </Text>
        </View>
      }
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onLayout={[Function]}
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "#757575",
                "fontSize": 16,
              }
            }
          >
            No Cities Added! Please Add City!
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
